pynumaflow/function/server.py

```diff
@@ -131,35 +138,72 @@ def MapFn(
 
         return udfunction_pb2.DatumList(elements=datums)
 
-    def ReduceFn(
-        self, request_iterator: Iterator[Datum], context: NumaflowServicerContext
+    async def ReduceFn(
+        self, request_iterator: AsyncIterable[Datum], context: NumaflowServicerContext
     ) -> udfunction_pb2.DatumList:
         """
         Applies a reduce function to a datum stream.
         The pascal case function name comes from the generated udfunction_pb2_grpc.py file.
         """
-        key, start, end = None, None, None
+
+        start, end = None, None
         for metadata_key, metadata_value in context.invocation_metadata():
-            if metadata_key == DATUM_KEY:
-                key = metadata_value
-            elif metadata_key == WIN_START_TIME:
+            if metadata_key == WIN_START_TIME:
                 start = metadata_value
             elif metadata_key == WIN_END_TIME:
                 end = metadata_value
-        if not (key or start or end):
+        if not (start or end):
             raise ValueError(
                 f"Expected to have all key/window_start_time/window_end_time; "
-                f"got key: {key}, start: {start}, end: {end}."
+                f"got start: {start}, end: {end}."
             )
 
         start_dt = datetime.fromtimestamp(int(start) / 1e3, timezone.utc)
         end_dt = datetime.fromtimestamp(int(end) / 1e3, timezone.utc)
         interval_window = IntervalWindow(start=start_dt, end=end_dt)
 
+        response_task = asyncio.create_task(
+            self.__async_reduce_handler(interval_window, request_iterator)
+        )
+
+        await response_task
+        return response_task.result()
+
+    async def __async_reduce_handler(self, interval_window, request_iterator: AsyncIterable[Datum]):
+        callable_dict = {}
+        # iterate through all the values
+        async for d in request_iterator:
+            key = d.key
+            rs = None
+            if key in callable_dict.keys():
```

Comment:
```suggestion
            if key in callable_dict:
```
If callable_dict is of type dict, this is much faster, but only if values of the callable_dict do not change in between the loop

Reply:
Is there an alternative thats faster? 
yes the keys do change with in the loop.

Reply:
I think let's try with the suggestion. `keys()` basically converts this to a tuple and returns a view of the keys. So it is O(N) operation. The suggestion on the other hand is a hash look up, so is O(1).
Can you see if that works?